### PR TITLE
DreamBooth support

### PR DESCRIPTION
**Pull request by filipstrand**
Very WIP at the moment...

**Updated: 2024-11-02** 
Some notes: 

Training does seem to work with 4bit on my machines with limited (32GB) memory. 

Even after just 200 gradient steps, we can at least produce very similar images to the [training dataset](https://github.com/ml-explore/mlx-examples/tree/main/flux#training-example)
test0000200_image

This result trained all single_transformer_blocks like so:

```python
    def set_lora_layer(self) -> None:
        for i in range(0, 38):
            block = self.transformer.single_transformer_blocks[i]
            block.proj_out = LoRALinear.from_linear(linear=block.proj_out)
            block.proj_mlp = LoRALinear.from_linear(linear=block.proj_mlp)
            block.attn.to_q = LoRALinear.from_linear(linear=block.attn.to_q)
            block.attn.to_k = LoRALinear.from_linear(linear=block.attn.to_k)
            block.attn.to_v = LoRALinear.from_linear(linear=block.attn.to_v)
```

I could not get good results when only training e.g the final single_transformer_blocks[37] as I think was done in [this example](https://huggingface.co/TheLastBen/The_Hound). This set me back some time thinking that something was wrong with the loss or similar. Now I believe loss is fine, and that there was simply not enough capacity for the model to fit the training examples when simply training single_transformer_blocks[37]

Image augmentations still to be done...

More investigations and example runs also needed. 

**Updated: 2024-12-12** 

Initial training code mostly finished. We now have full resumable and deterministic training. 
There is still improvements to be made (like better image augmentation pre-processing, start with cmd line args in addition to the config file etc.), but could be done after initial release. 
What remains (in progress) is a better and unified handling of the LoRA weights. 

**Updated: 2024-12-15**
Nearing completion. Required quite a big rewrite of the weight handling to handle LoRAs, but now seem to work as intended. Have added a few tests that ensure outputs are not substantially different compared to before. Small tweaks and cleanups left

### CODE REVIEW

**Comment by anthonywu** [comment]
been constrained on time this week but I hope to give this a review and test the dreambooth training myself

**Comment by filipstrand** [comment]
@anthonywu No worries :) This is still very much WIP at this stage (hardcoded paths, arbitrary parameter settings, etc) so be mindful.  So far I've only tried it with `dev`, 512x512 resolution and in 4bit since it takes quite a while for these to train on my machines (~ 10h for around 600 gradient steps). 

Depending on the machine you have, I think there are big difference in training speed. If you do try it out, it would be fun to see your numbers, e.g how long it takes to perform 1 gradient steps with the current setup:  

```
Epoch 5/50:  48%|████▊     | 48/100 [1:08:30<1:18:40, 90.77s/it, loss=0.6]

You can see the iteration speed here as:  90.77s per step
```

In my case, I am hitting swap memory when training with the current setup (all singe transformer block layers), and the GPU is not fully utilised (you can barely hear the fans going). Training less amounts of parameters helps with this, but then I think I get the problem described above of using too few parameters (though not fully sure about this, maybe just have to train for longer...).

**Comment by anthonywu** [comment]
Due to current time constraints I won't be able to help review this PR. Looks like you're doing a long test cycle and diligence prior to release so I'm sure you'll be fine here! Thanks for this big feature and looking forward to seeing this shipped!

**Comment by filipstrand** [comment]
> Due to current time constraints I won't be able to help review this PR. Looks like you're doing a long test cycle and diligence prior to release so I'm sure you'll be fine here! Thanks for this big feature and looking forward to seeing this shipped!

Thanks Anthony! No worries at all, but thanks for letting me know :). This PR starts to feel ready enough for initial release except for a few things... It admittedly took much longer than I initially planed (and required a rewrite of the weight-handling logic to unify how we load LoRAs in all cases - but now it is starting to look good and I have confidence in the tests). If I accidentally broke some existing behaviour, I am sure someone would let me know :)  

I'm planning on releasing this feature before Christmas and I'm sure there will be feedback on things over time that can be improved (such as input flags in addition to config file etc.). The best strategy here is probably to release it as v1 and iterate from there.

**Comment by gmorain** [comment]
Hi @filipstrand I was just following the discussion in this PR as I did not have enough spare time to contribute substantially to the effort (I am probably not competent enough for the coding part, but an additional tester would help I guess). Happy if you can achieve you goal to release the feature before Christmas since I will probably have more time to participate in testing the feature (Macbook Pro M1, 16 GB, macOS Ventura 13.7.1 here). Just taking also the opportunity to thank you for the work you are putting into this!

**Comment by filipstrand** [comment]
Hi @gmorain! Very thankful for your willingness to help out, and more testing is always welcome :) 

Since I've made quite a big change to how we load and handle the LoRAs, just to make sure nothing is broken, if you have your own examples of custom LoRAs downloaded from the internet, it would be great to hear if they produce the same image with this `dreambooth` branch compared to `main` or the 0.4.1 release (given the exact same settings like prompt, seed etc). I have added some automated tests for this, but you can never be fully sure (and since LoRA formats can differ a bit, there is a chance that some other LoRAs might work different than the ones I have tried). If you have the time, example of loading multiple LoRAs would also be great to confirm. 

Would also be very interested to hear if you would be able to achieve a training run on 512x512 resolution on 16GB of memory (my guess is not, since my 32GB machines can barley do it with the following config:

```
{
  "model": "dev",
  "seed": 42,
  "steps": 20,
  "guidance": 3.0,
  "quantize": 4,
  "width": 512,
  "height": 512,
  "training_loop": {
    "num_epochs": 100,
    "batch_size": 1
  },
  "optimizer": {
    "name": "AdamW",
    "learning_rate": 1e-4
  },
  "save": {
    "output_path": "~/Desktop/train",
    "checkpoint_frequency": 10
  },
  "instrumentation": {
    "plot_frequency": 1,
    "generate_image_frequency": 25,
    "validation_prompt": "photo of sks dog"
  },
  "lora_layers": {
    "single_transformer_blocks" : {
        "block_range": {
            "start": 0,
            "end": 38
        },
        "layer_types": [
            "attn.to_q",
            "attn.to_k",
            "attn.to_v"
        ],
        "lora_rank": 4
        }
    },
```

**Comment by filipstrand** [comment]
@azrahello I think this is fixed now in the latest main. If you still see the same slowdown, please let me know :)

**Comment by azrahello** [comment]
> @azrahello I think this is fixed now in the latest main. If you still see the same slowdown, please let me know :)

I’m doing it now! :D Python is using 45 GB, it starts at 30s/it, and after 10 steps, it seems to stabilize at around 14. I did the pull after your comment.

**Comment by azrahello** [comment]
I made some changes to the plot file to make the graph a bit more verbose. I was playing around with the IDE, and the file seemed ‘approachable’ for running some tests.
[0000300_loss.pdf](https://github.com/user-attachments/files/18222761/0000300_loss.pdf)

**Comment by filipstrand** [comment]
@azrahello Nice! If you want, you can publish a PR with your changes when you feel happy with them. I am planning on merging this branch to main today and do a v.0.5.0 release, but over the coming time there are some obvious improvements that can be made to the training part overall, and improving the visualisation can certainly be one of them :). 

Also, thanks for your testing efforts for this PR!

**Comment by gmorain** [comment]
> Hi @gmorain! Very thankful for your willingness to help out, and more testing is always welcome :)
> 
> Since I've made quite a big change to how we load and handle the LoRAs, just to make sure nothing is broken, if you have your own examples of custom LoRAs downloaded from the internet, it would be great to hear if they produce the same image with this `dreambooth` branch compared to `main` or the 0.4.1 release (given the exact same settings like prompt, seed etc). I have added some automated tests for this, but you can never be fully sure (and since LoRA formats can differ a bit, there is a chance that some other LoRAs might work different than the ones I have tried). If you have the time, example of loading multiple LoRAs would also be great to confirm.
> 
> Would also be very interested to hear if you would be able to achieve a training run on 512x512 resolution on 16GB of memory (my guess is not, since my 32GB machines can barley do it with the following config:
> 
> ```
> {
>   "model": "dev",
>   "seed": 42,
>   "steps": 20,
>   "guidance": 3.0,
>   "quantize": 4,
>   "width": 512,
>   "height": 512,
>   "training_loop": {
>     "num_epochs": 100,
>     "batch_size": 1
>   },
>   "optimizer": {
>     "name": "AdamW",
>     "learning_rate": 1e-4
>   },
>   "save": {
>     "output_path": "~/Desktop/train",
>     "checkpoint_frequency": 10
>   },
>   "instrumentation": {
>     "plot_frequency": 1,
>     "generate_image_frequency": 25,
>     "validation_prompt": "photo of sks dog"
>   },
>   "lora_layers": {
>     "single_transformer_blocks" : {
>         "block_range": {
>             "start": 0,
>             "end": 38
>         },
>         "layer_types": [
>             "attn.to_q",
>             "attn.to_k",
>             "attn.to_v"
>         ],
>         "lora_rank": 4
>         }
>     },
> ```

Sorry @filipstrand for the false promise, I was too tired and worked too much on another computer recently that actually had 16 GB of RAM. My M1 Macbook actually has 32 GB of RAM. But I am currently nevertheless testing the v0.5.0 you pushed yesterday with the settings above and the dog sample images to see if everything works fine on my setup too.
